### ISSUE TITLE
mig-controller may need changes to work with OCP 4.4

### ISSUE THREAD

**Issue by jmontleon** [konveyor/mig-controller#432]
@Danil-Grigorev is seeing an error when trying to use mig-controller on OCP 4.4.

In the error below we have:
```
""error":"no matches for kind \"Deployment\" in version \"apps/v1beta1\"
```

Coming from:
https://github.com/konveyor/mig-controller/blob/master/pkg/controller/migplan/migplan_controller.go#L311

Which is probably being thrown here:
https://github.com/konveyor/mig-controller/blob/93bda35aabb93cec6b3aaac9a4031f67b874f0be/pkg/apis/migration/v1alpha1/migcluster_types.go#L184

apps/v1beta1 Deployments no longer exists in OpenShift 4.4 and we must use apps/v1 Deployments.

I don't know if this approach will work for the controller and whether this is cross cluster so source and destination versions need to be taken into account, but in the operator I took the approach of checking the Kubernetes minor version and if old enough using v1beta1 else v1. It was also necessary to do this for daemonsets
https://github.com/konveyor/mig-operator/pull/232/files
https://github.com/konveyor/mig-operator/pull/233/files

Full error:
```
{"level":"error","ts":1582726073.6475673,"logger":"plan|bjp24","msg":"","error":"no matches for kind \"Deployment\" in version \"apps/v1beta1\"","stacktrace":"github.com/konveyor/mig-controller/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/konveyor/mig-controller/pkg/logging.Logger.Error\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/logging/logger.go:75\ngithub.com/konveyor/mig-controller/pkg/logging.Logger.Trace\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/logging/logger.go:81\ngithub.com/konveyor/mig-controller/pkg/controller/migplan.(*ReconcileMigPlan).ensureClosed\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/migplan_controller.go:311\ngithub.com/konveyor/mig-controller/pkg/controller/migplan.(*ReconcileMigPlan).handleClosed\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/migplan_controller.go:294\ngithub.com/konveyor/mig-controller/pkg/controller/migplan.(*ReconcileMigPlan).Reconcile\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/migplan_controller.go:196\ngithub.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```

**Comment by jortel** [comment]
The proposed solution would be good for the controller as well.
I'm thinking that the cluster controller could use the `/version` API:
```
url -k --header "Authorization: Bearer $token" https://host:6443/version
{
  "major": "1",
  "minor": "16+",
  "gitVersion": "v1.16.2",
  "gitCommit": "94e669a",
  "gitTreeState": "clean",
  "buildDate": "2020-02-03T23:11:39Z",
  "goVersion": "go1.12.12",
  "compiler": "gc",
  "platform": "linux/amd64"
}
```
and store the version (major, minor) in the MigCluster spec.
All of the controllers could use the information to determine which GVKs to use when necessary.
When this is available, we should use this approach for `CronJob` instead of the _unstructured_ approach currently used.

**Comment by sseago** [comment]
Are there two separate situations to deal with here?
1) Make sure that any resources created or referenced by the controller use the proper version (and do we have to determine cluster version first? i.e. create one resource for OCP3.7 and a different for OCP4.5, etc.)
2) For migrations, we may have to *change* the version when restoring a Deployment, etc.

**Comment by jmontleon** [comment]
There's nothing substantially different between v1beta1 and v1 deployments.

**Comment by jmontleon** [comment]
There's nothing substantially different between v1beta1 and v1 deployments.

matchLabels are required in v1 and were not in v1beta1, though they were acceptable.
https://github.com/konveyor/mig-operator/pull/232/files#diff-97f72c03b7703675f1cff5f5606488ceL18

If creating deployments make sure these are included and all you have to do is adjust the version.

**Comment by sseago** [comment]
Is v1 available for all OCP versions we support, or would we have to check OCP version for the target cluster (src, dest, or controller cluster) and modify the APIversion appropriately?

**Comment by jmontleon** [comment]
It is not available for 3.7. That's why we were using v1beta1. Now v1beta1 is not available in 4.4 so we can't overlap everything with v1beta1.

**Comment by Danil-Grigorev** [comment]
It is a perfect opportunity to include creation of discovery client to the default `MigCluster` functionality.

https://godoc.org/k8s.io/client-go/discovery#DiscoveryClient.ServerVersion
